lib/debug.ml

```diff
@@ -151,17 +151,23 @@ let set_facility f = Mutex.execute facility_m (fun () -> facility := f)
 let get_facility () = Mutex.execute facility_m (fun () -> !facility)
 
 let output_log brand level priority s =
-  if not(is_disabled brand level) then begin
     let msg = format false brand priority s in
     if !print_debug
     then Printf.printf "%s\n%!" (format true brand priority s);
-
     Syslog.log (get_facility ()) level msg
+
+(* CA-326244: This function takes a lock to check if the debug level is disabled,
+   therefore it shouldn't be called from exception hooks, finalizers, etc. that
+   can get invoked from any place in the code, that might result in a deadlock in this module in
+   case it's invoked while we already hold the loglevel_m mutex *)
+let output_log_if_enabled brand level priority s =
```

Comment:
Does `is_disabled()` really need a lock? 
```
let is_disabled brand level =
  Mutex.execute loglevel_m (fun () ->
      Syslog.is_masked ~threshold:!loglevel level ||
      Hashtbl.mem logging_disabled_for (brand, level)
    )
```
I understand that it reads two values and thus the lock guarantees consistency but it seems to introduce more problems than it solves.

Reply:
I will impelemnt @edwintorok 's suggestion of using a Set reference, which should be safe to read without a lock.
I asked @edwintorok the same question and he said it might be dangerous to read the hash table while it's being modified.